src/shared/analytics/mixpanel.ts

```diff
@@ -0,0 +1,264 @@
+        ...this.baseProperties,
+        time: Date.now() / 1000,
+        $insert_id: crypto.randomUUID(),
+        $device_id: this.deviceId,
+        distinct_id: this.userId ?? `$device:${this.deviceId}`,
+        token: this.token,
+        ...values,
+        ...(this.userId
+          ? {
+              $user_id: this.userId,
+              // duplicated user_id is a current product requirement
+              user_id: this.userId,
+            }
+          : null),
+      },
+    };
+
+    log(Loglevel.info, payload.event);
+    logTable(Loglevel.info, payload.properties);
+
+    return this.sendEvent(url.toString(), { json: [payload] });
+  }
+
+  async engage(userProfileProperties: Record<string, unknown>) {
+    const url = new URL('/engage', this.url);
+    if (this.debugMode) {
+      url.searchParams.append('verbose', '1');
+    }
+    url.searchParams.append('ip', '1');
+    if (!this.userId) {
+      // eslint-disable-next-line no-console
+      console.warn('Must not call engage() method without a user id');
+      return;
+    }
+    const payload = {
+      $device_id: this.deviceId,
+      $distinct_id: this.userId,
+      $user_id: this.userId,
+      token: this.token,
+      $set: userProfileProperties,
+    };
+
+    logTable(Loglevel.info, payload);
+
+    return this.sendEvent(url.toString(), { json: [payload] });
+  }
+
+  async identify(
+    userId: string,
+    userProfileProperties: Record<string, unknown>
+  ) {
+    await this.ready();
+    this.userId = userId;
+    const $anon_distinct_id = `$device:${this.deviceId}`;
+    return Promise.all([
+      // TODO: "$identify" track event is not necessary?
+      this.track('$identify', { $anon_distinct_id }),
+      this.engage(userProfileProperties),
+    ]);
+  }
+
+  async sendEvent(url: string, options: Options) {
+    if (this.token != null && this.sendRequestsOverTheNetwork) {
+      return ky.post(url, options);
+    }
+  }
+
+  reset() {
+    // we're not resetting "super" properties because in our case
+    // they're not user-specific
+    this.userId = undefined;
+  }
+}
+
+const mixpanelApi = new MixpanelApi({
+  token: mixPanelToken ?? null,
+  resolveDeviceId: () => deviceIdStore.getSavedState(),
+  debugMode: process.env.NODE_ENV !== 'production',
+  sendRequestsOverTheNetwork: process.env.NODE_ENV === 'production',
+});
+
+Object.assign(globalThis, { mixpanelApi });
+
+mixpanelApi.setBaseProperties({
+  origin: globalThis.location.origin,
+  app_version: version,
+});
+
+export async function mixPanelTrack(
+  account: Account,
+  event: string,
+  values: Record<string, unknown>
+) {
+  const userId = account.getUser()?.id;
+  mixpanelApi.track(event, {
+    ...values,
+    ...(userId ? { user_id: userId } : null),
+  });
+}
+
+export async function mixpanelIdentify(account: Account) {
+  const userId = account.getUser()?.id;
+  if (!userId) {
+    return;
+  }
+  try {
+    const userProfileProperties = await getBaseMixpanelParams(account);
```

Comment:
Did we discuss ability to turn off analytics by demand? What if somebody doesn't want to share this?

Reply:
No sensitive data is supposed to be here. We don't send addresses to analytics